generator/genutils/utils.go

```diff
@@ -0,0 +1,138 @@
+	"sigs.k8s.io/controller-tools/pkg/crd"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+const (
+	patchStrategyTagKey = "patchStrategy"
+	patchMergeKeyTagKey = "patchMergeKey"
+	// MergePatchStrategy is the name of the Merge patch strategy
+	MergePatchStrategy = "merge"
+	// ReplacePatchStrategy is the name of the Replace patch strategy
+	ReplacePatchStrategy = "replace"
+)
+
+// ContainsPatchStrategy reads the field tags to check whether the given patch strategy is defined
+func ContainsPatchStrategy(field *markers.FieldInfo, strategy string) bool {
+	patchStrategy := field.Tag.Get(patchStrategyTagKey)
+	if patchStrategy == "" {
+		return false
+	}
+
+	for _, s := range strings.Split(patchStrategy, ",") {
+		if s == strategy {
+			return true
+		}
+	}
+	return false
+}
+
+// GetPatchMergeKey reads the field tags to retrieve the patch merge key. It returns nil if no such key is defined
+func GetPatchMergeKey(field *markers.FieldInfo) string {
+	return field.Tag.Get(patchMergeKeyTagKey)
+}
+
+// Visitor is the type of a function that visits one level of Json schema
+type Visitor func(schema *apiext.JSONSchemaProps) (newVisitor Visitor, stop bool)
+
+type visitorStruct struct {
+	VisitFunc Visitor
+}
+
+func (v visitorStruct) Visit(schema *apiext.JSONSchemaProps) crd.SchemaVisitor {
+	newVisitor, stop := v.VisitFunc(schema)
+	if stop {
+		return nil
+	}
+
+	if newVisitor == nil {
+		return v
+	}
+	return visitorStruct{newVisitor}
+}
+
+// EditJSONSchema allows editing a Json Schema by applying the visitor function to each level of the schema.
+func EditJSONSchema(jsonSchema *apiext.JSONSchemaProps, visitor Visitor) {
+	crd.EditSchema(jsonSchema, visitorStruct{visitor})
+}
+
+// AddUnionOneOfConstraints adds oneOf constraints in the given Json schema for all the unions, based on the authorized values of the union members
+func AddUnionOneOfConstraints(jsonSchema *apiext.JSONSchemaProps, unionDiscriminators []markers.FieldInfo, removeDiscriminators bool, fieldsToSkip ...string) {
+	mainVisitor := func(schema *apiext.JSONSchemaProps) (newVisitor Visitor, stop bool) {
+		if schema == nil {
+			return
+		}
+		if schema.Type != "object" {
+			return
+		}
+
+		if len(schema.Properties) > 0 {
+		discriminatorLoop:
+			for _, discriminator := range unionDiscriminators {
+				discriminatorPropertyName := strcase.ToLowerCamel(discriminator.Name)
+				if discriminatorProp, found := schema.Properties[discriminatorPropertyName]; found &&
+					discriminatorProp.Enum != nil &&
+					len(discriminatorProp.Enum) != 0 {
+					oneOf := []apiext.JSONSchemaProps{}
+					newEnumValues := []apiext.JSON{}
+				enumValueLoop:
+					for _, enumValue := range discriminatorProp.Enum {
+						fieldName := strings.Trim(string(enumValue.Raw), `"`)
+						unionMemberProperty := strcase.ToLowerCamel(fieldName)
+						if _, foundUnionMember := schema.Properties[unionMemberProperty]; !foundUnionMember {
+							continue discriminatorLoop
```

Comment:
The discriminator enum contains values that do not correspond to any enum field. So we just skip the `oneOf` creation here since this union definition seems invalid.
TBF we should be more strict and implement real union validity checks, but I didn't do it for now.
I'll add a comment in the code to explain the logic here at least.

Reply:
Fixed in commit https://github.com/devfile/api/pull/133/commits/3465390b28039c68fb26c1db19b1429bc55dc2c1